src/Microsoft.Data.Analysis/DataFrame.IO.cs

```diff
@@ -214,87 +193,189 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
-                    columns.Add(CreateColumn(kind, columnNames, i));
-                }
+            if (rowline == 0)
+            {
+                throw new FormatException(Strings.EmptyFile);
+            }
 
-                DataFrame ret = new DataFrame(columns);
-                line = null;
-                streamReader.DiscardBufferedData();
-                streamReader.BaseStream.Seek(streamStart, SeekOrigin.Begin);
+            columns = new List<DataFrameColumn>(numberOfColumns);
+            // Guesses types or looks up dataTypes and adds columns.
+            for (int i = 0; i < numberOfColumns; ++i)
+            {
+                Type kind = dataTypes == null ? GuessKind(i, linesForGuessType) : dataTypes[i];
+                columns.Add(CreateColumn(kind, columnNames, i));
+            }
+
+            DataFrame ret = new DataFrame(columns);
+            line = null;
 
-                // Fills values.
-                line = streamReader.ReadLine();
-                rowline = 0;
-                while (line != null && (numberOfRowsToRead == -1 || rowline < numberOfRowsToRead))
+            // Fill values.
+            enumerator.Reset();
+            rowline = 0;
+            while (enumerator.MoveNext() && (numberOfRowsToRead == -1 || rowline < numberOfRowsToRead))
+            {
+                line = enumerator.Current;
+                var spl = line.Split(separator);
+                if (header && rowline == 0)
                 {
-                    var spl = line.Split(separator);
-                    if (header && rowline == 0)
-                    {
-                        // Skips.
-                    }
-                    else
-                    {
-                        ret.Append(spl, inPlace: true);
-                    }
-                    ++rowline;
-                    line = streamReader.ReadLine();
+                    // Skips.
+                }
+                else
+                {
+                    ret.Append(spl, inPlace: true);
                 }
+                ++rowline;
+            }
 
-                if (addIndexColumn)
+            if (addIndexColumn)
+            {
+                PrimitiveDataFrameColumn<int> indexColumn = new PrimitiveDataFrameColumn<int>("IndexColumn", columns[0].Length);
+                for (int i = 0; i < columns[0].Length; i++)
                 {
-                    PrimitiveDataFrameColumn<int> indexColumn = new PrimitiveDataFrameColumn<int>("IndexColumn", columns[0].Length);
-                    for (int i = 0; i < columns[0].Length; i++)
-                    {
-                        indexColumn[i] = i;
-                    }
-                    columns.Insert(0, indexColumn);
+                    indexColumn[i] = i;
                 }
-                return ret;
+                columns.Insert(0, indexColumn);
+            }
+            return ret;
+        }
+
+        private struct CsvLines : IEnumerable<string>
```

Comment:
Does this need to be a `struct`? You are going to box it as soon as you pass it into `ReadCsvLinesIntoDataFrame(IEnumerable<string> lines` anyway.

If you really wanted to save the single alloc, you could make `ReadCsvLinesIntoDataFrame(IEnumerable<string> lines` be generic `where T : IEnumerable<string>`.

Reply:
Oh, I didn't realize that value type -> interfaces also does boxing. Just looked it up. Nice! Fixed it now. Just made it a class.